pkg/pipelinerunmetrics/metrics.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerunmetrics
+
+import (
+	"context"
+	"sync"
+
+	"github.com/tektoncd/chains/pkg/chains"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	sgCount = stats.Float64(chains.PipelineRunSignedName,
+		chains.PipelineRunSignedDesc,
+		stats.UnitDimensionless)
+
+	sgCountView *view.View
+
+	plCount = stats.Float64(chains.PipelineRunUploadedName,
+		chains.PipelineRunUploadedDesc,
+		stats.UnitDimensionless)
+
+	plCountView *view.View
+
+	stCount = stats.Float64(chains.PipelineRunStoredName,
+		chains.PipelineRunStoredDesc,
+		stats.UnitDimensionless)
+
+	stCountView *view.View
+
+	mrCount = stats.Float64(chains.PipelineRunMarkedName,
+		chains.PipelineRunMarkedDesc,
+		stats.UnitDimensionless)
+
+	mrCountView *view.View
+)
+
+// Recorder holds keys for Tekton metrics
+type Recorder struct {
+	initialized bool
+}
+
+// We cannot register the view multiple times, so NewRecorder lazily
+// initializes this singleton and returns the same recorder across any
+// subsequent invocations.
+var (
+	once           sync.Once
+	r              *Recorder
+	errRegistering error
+)
+
+// NewRecorder creates a new metrics recorder instance
+// to log the PipelineRun related metrics
+func NewRecorder(ctx context.Context) (*Recorder, error) {
+	logger := logging.FromContext(ctx)
+	once.Do(func() {
```

Comment:
Instead of using this, why not just execute this in `init`? You won't have access to the context, but it's probably fine to not log messages here. If `viewRegister` fails, should that be a fatal error? Doing this in `init` probably means you don't need the `initialized` attribute in the `Recorder`.

Reply:
With init, wouldn't viewRegister  registered at initialization stage by default? in our case we are also calling newrecorder to register client context as well.

Reply:
The main benefit I see with the current approach is that it allows a context to be passed in which is then used for logging. This wouldn't be possible if the Recorder was added during `init`. Let's leave it as is then.

Reply:
ok